src/main/scala/org/broadinstitute/hail/asm4s/Code.scala

```diff
@@ -349,4 +508,44 @@ class CodeObject[T <: AnyRef](val lhs: Code[T])(implicit tct: ClassTag[T], tti:
   def invoke[A1, A2, S](method: String, a1: Code[A1], a2: Code[A2])(implicit a1ct: ClassTag[A1], a2ct: ClassTag[A2],
                                                                     sct: ClassTag[S]): Code[S] =
     invoke[S](method, Array[Class[_]](a1ct.runtimeClass, a2ct.runtimeClass), Array[Code[_]](a1, a2))
+
+  def invoke[A1, A2, A3, S](method: String, a1: Code[A1], a2: Code[A2], a3: Code[A3])
+    (implicit a1ct: ClassTag[A1], a2ct: ClassTag[A2], a3ct: ClassTag[A2], sct: ClassTag[S]): Code[S] =
+    invoke[S](method, Array[Class[_]](a1ct.runtimeClass, a2ct.runtimeClass, a3ct.runtimeClass), Array[Code[_]](a1, a2, a3))
+
+  def ifNull[T](cnullcase: Code[T], cnonnullcase: Code[T]): Code[T] =
+    new Code[T] {
+      def emit(il: java.util.Collection[AbstractInsnNode]): Unit = {
+        val lnull = new LabelNode
+        val lafter = new LabelNode
+        lhs.emit(il)
+        il.add(new JumpInsnNode(IFNULL, lnull))
+        cnonnullcase.emit(il)
+        il.add(new JumpInsnNode(GOTO, lafter))
+        il.add(lnull)
+        cnullcase.emit(il)
+        // fall through
+        il.add(lafter)
+      }
+    }
+
+  def mapNull[U >: Null](cnonnullcase: Code[U]): Code[U] =
+    ifNull[U](Code._null[U], cnonnullcase)
+
+  def mapNull[U >: Null](cnonnullcase: Code[T] => Code[U]): CM[Code[U]] = for (
+    (stx, x) <- CM.memoize(lhs)
+  ) yield Code(stx, cnonnullcase(x))
+
+
+  // def mapNull[U >: Null](cnonnullcase: Code[T] => Code[U]): Code[U] =
```

Comment:
This doesn't work unless `cnonnullcase` uses the argument before editing the stack. It'd be nice to have `Code` somehow track the changes in the stack so that we can access stack variables via `swap`s and such. Maybe it doesn't matter though. I'm not sure of the performance comparison between stack and register accesses.